ansible/roles/test/files/ptftests/sonic.py

```diff
@@ -0,0 +1,618 @@
+from ptf.testutils import *
+from ptf.dataplane import match_exp_pkt
+import datetime
+import _strptime  # workaround python bug ref: https://stackoverflow.com/a/22476843/2514803
+import time
+from ptf.mask import Mask
+import ptf.packet as scapy
+from six.moves import _thread as thread
+import threading
+import os
+import sys
+import json
+import re
+from collections import defaultdict
+import paramiko
+import pickle
+from operator import itemgetter
+import scapy.all as scapyall
+import ast
+import socket
+
+import host_device
+
+class Sonic(host_device.HostDevice):
+    DEBUG = False
+    # unit: second
+    SSH_CMD_TIMEOUT = 10
+    def __init__(self, ip, queue, test_params, log_cb=None, login='admin', password='password'):
+        self.ip = ip
+        self.queue = queue
+        self.log_cb = log_cb
+        self.login = login
+        self.password = password
+        self.conn = None
+        self.v4_routes = list(ast.literal_eval(test_params['vlan_ip_range']).values())
+        self.v4_routes.append(test_params['lo_prefix'])
+        self.v6_routes = [test_params['lo_v6_prefix']]
+        self.fails = set()
+        self.info = set()
+        self.min_bgp_gr_timeout = int(test_params['min_bgp_gr_timeout'])
+        self.reboot_type = test_params['reboot_type']
+        self.bgp_v4_v6_time_diff = test_params['bgp_v4_v6_time_diff']
+        self.port_channel_intf_idx = test_params['port_channel_intf_idx']
+        self.port_channel_last_lacp_pdu_time = None
+        self.port_channel_last_lacp_pdu_time_lock = threading.Lock()
+        self.lacp_pdu_time_on_down = list()
+        self.lacp_pdu_time_on_up = list()
+
+    def log(self, msg):
+        if self.log_cb is not None:
+            self.log_cb('SSH thread VM={}: {}'.format(self.ip, msg))
+
+    def connect(self):
+        if self.conn is None:
+            self.conn = paramiko.SSHClient()
+            self.conn.set_missing_host_key_policy(paramiko.AutoAddPolicy())
+            self.conn.connect(self.ip, username=self.login, password=self.password, allow_agent=False, look_for_keys=False)
+
+        self.show_lacp_command = self.parse_supported_show_lacp_command()
+        self.show_ip_bgp_command = self.parse_supported_bgp_neighbor_command()
+        self.show_ipv6_bgp_command = self.parse_supported_bgp_neighbor_command(v4=False)
+        return self.conn
+
+    def do_cmd(self, cmd):
+        while True:
+            try:
+                stdin, stdout, stderr = self.conn.exec_command(cmd, timeout=Sonic.SSH_CMD_TIMEOUT)
+                return stdout.read()
+            except socket.timeout as e:
+                self.log("Timeout when running command: {}".format(cmd))
+                return ""
+            except paramiko.SSHException as e:
+                # Possibly caused by https://github.com/paramiko/paramiko/issues/822
+                # Disconnect and reconnect as a possible workaround?
+                self.disconnect()
+                self.connect()
```

Comment:
It's in a `while True` loop for this purpose, although there's currently no limit on the number of retries. That should probably be added so that it doesn't get stuck here.

Reply:
Limit of 5 attempts set.